### PR TITLE
Stochastic surface weight (Uniform(10,40) per batch)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
Instead of a fixed adaptive surf_weight per epoch, randomly sample surf_weight from Uniform(10,40) each training batch. This turns the model into an implicit ensemble over different vol/surf tradeoffs, finding Pareto-optimal parameters. At evaluation, use the mean value (25). The randomization acts as a regularizer similar to dropout.

## Instructions
In `structured_split/structured_train.py`, in the training loop where `surf_weight` is computed:

1. Replace the adaptive surf_weight computation with stochastic sampling during training:
```python
if model.training:
    surf_weight = torch.empty(1, device=device).uniform_(10.0, 40.0).item()
else:
    surf_weight = 25.0
```

2. Keep the adaptive surf_weight code but override it. The existing `sw_ema` tracking can remain for logging.

3. For the tandem boost, still apply `surf_weight *= 1.5` for tandem samples after the random sampling.

Run with: `--wandb_name "frieren/stoch-sw" --wandb_group stochastic-surf-weight --agent frieren`

## Baseline
- val/loss: **2.3396**
- val_in_dist/mae_surf_p: 21.49
- val_ood_cond/mae_surf_p: 22.68
- val_ood_re/mae_surf_p: 31.60
- val_tandem_transfer/mae_surf_p: 44.28

---

## Results

**W&B run:** `vyt93fzc` (~78 epochs, 29.1 min, timed out)

| Metric | Baseline | This run | Δ |
|--------|----------|----------|---|
| val/loss | 2.3396 | **2.4847** | +6.2% |
| val_in_dist/mae_surf_p | 21.49 | **21.25** | **-1.1%** |
| val_ood_cond/mae_surf_p | 22.68 | **23.99** | +5.8% |
| val_ood_re/mae_surf_p | 31.60 | **32.55** | +3.0% |
| val_tandem_transfer/mae_surf_p | 44.28 | **45.82** | +3.5% |

Surface MAE (Ux, Uy, p):
- val_in_dist: (0.290, 0.176, 21.25), mean = 7.24
- val_ood_cond: (0.294, 0.194, 23.99), mean = 8.16
- val_tandem_transfer: (0.691, 0.360, 45.82), mean = 15.62

Volume MAE:
- val_in_dist: (1.889, 0.671, 37.36), mean = 13.31
- val_ood_cond: (1.550, 0.570, 28.86), mean = 10.33
- val_tandem_transfer: (2.776, 1.309, 55.70), mean = 19.93

### What happened
Clearly negative result — val/loss is +6.2% worse. Volume MAE worsened significantly across all splits (in_dist vol: 12.45→13.31; ood_cond vol: 9.30→10.33), and surface pressure regressed in most OOD splits. Only in-dist surf_p showed a tiny improvement (-1.1%).

**Why it hurt:** Stochastic surf_weight introduces severe gradient variance — when surf_weight=10 the vol loss dominates, when surf_weight=40 the surface loss dominates. Rather than finding a Pareto-optimal solution, the model gets conflicting gradient signals every batch, making it harder to converge. The implicit ensemble effect is real in theory but doesn't outweigh the optimization instability at this training horizon (~78 epochs).

The adaptive loss-ratio baseline (2.3396) is already a principled approach to setting surf_weight — it auto-adjusts to keep vol and surf losses balanced. Random sampling doesn't improve on this.

### Suggested follow-ups
- Try a narrower stochastic range (e.g., Uniform(15, 30)) to reduce variance while keeping the ensemble effect.
- Try stochastic surf_weight only in early training, switching to fixed adaptive weight in later epochs.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: **Arm A Kill + Arm B Wave-Best Update**

**Arm A (`aoetlx9b`, α=1.0) — KILLED**
W&B shows Arm A entered state `killed` at step 65,161 (~4-GPU EP5.93). This is NOT a kill gate trigger — Arm A cleared both the EP5 gate (≤9.0%, actual 6.9162%) and the EP10 gate (≤8.0%, actual 6.9162%). The kill appears to be a timeout or external event.

**I need you to explain what happened to Arm A.** Please post:
1. The exact termination reason (SENPAI_TIMEOUT hit? Pod evicted? Manual kill?)
2. Whether you have the EP6 checkpoint or if training is recoverable
3. Final W&B summary metrics for Arm A

**Arm B (`g18f7jm1`, α=0.5) — RUNNING, CURRENT WAVE BEST**
Arm B is still running at step ~65,056 (4-GPU EP5.93). Its current best is **6.7438%** at 4-GPU EP5 — the best result in this entire wave, 0.214pp above SOTA val_best (6.5281%).

**Performance comparison (4-GPU epochs, 10,986 steps/epoch):**
| 4-GPU Epoch | Arm A (α=1.0) | Arm B (α=0.5) | Gap |
|-------------|---------------|---------------|-----|
| EP1 | 8.6951% | 8.6379% | −0.057pp |
| EP2 | 7.5078% | 7.4012% | −0.107pp |
| EP3 | 7.1901% | 7.0931% | −0.097pp |
| EP4 | 7.0836% | 6.8721% | −0.212pp |
| EP5 | 6.9162% | 6.7438% | −0.212pp |

The α=0.5 advantage is clear and growing. α=0.5 finds a more balanced equilibrium — it avoids over-amplifying the τ_z weight spread that α=1.0 exhibits, allowing cp and vol_p to continue improving alongside τ_z.

**Arm B: continue running, target EP20 gate (≤7.2%)**
Both EP5 (≤9.0%) and EP10 (≤8.0%) gates are already cleared. The EP20 gate is ≤7.2%; Arm B at EP5 = 6.7438% is well ahead of pace.

**At your next 4-GPU even epoch (EP6), please post:**
1. Current val_abupt
2. Per-channel weight breakdown from GradNorm (normalized weights for cp, τ_x, τ_y, τ_z at this epoch)
3. Clarification on Arm A termination

Do NOT launch any new arms until Arm B has been confirmed through EP10.
